### PR TITLE
[hbase10] Address #701 by mimicking the same locks from the HBase 0.9…

### PR DESCRIPTION
**Pull request by manolama**
…8 client in the HBase 10 client.

### CODE REVIEW

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java [hunk 1]
```diff
@@ -82,7 +84,7 @@
    * @See #CONNECTION_LOCK.
    */
   private static Connection connection = null;
-  private static final Object CONNECTION_LOCK = new Object();
+  //private static final Object CONNECTION_LOCK = new Object();
```
Comment (allanbank) [inline hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: Just delete the line. Don't comment out.

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java [hunk 2]
```diff
@@ -225,7 +224,7 @@ public void cleanup() throws DBException {
 
   public void getHTable(String table) throws IOException {
     final TableName tName = TableName.valueOf(table);
-    synchronized (CONNECTION_LOCK) {
+    synchronized (TABLE_LOCK) {
```
Comment (allanbank) [inline hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: #701 asked for this lock to be removed, if I am reading it correctly.

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java [hunk 3]
```diff
@@ -208,14 +210,11 @@ public void cleanup() throws DBException {
...
-        synchronized (CONNECTION_LOCK) {
-          if (connection != null) {
-            connection.close();
-            connection = null;
-          }
+      synchronized (THREAD_COUNT) {
+        int threadCount = THREAD_COUNT.decrementAndGet();
+        if (threadCount <= 0 && connection != null) {
+          connection.close();
+          connection = null;
```
Comment (allanbank) [inline hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: This synchronized block should not be needed. The closing of the connection is protected by the decrement of the atomic integer.

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java [hunk 4]
```diff
@@ -145,8 +147,8 @@ public void init() throws DBException {
     }
 
     try {
```
Comment (allanbank) [inline hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: Can we merge this try into the try after the "Terminate right now" comment?

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java [hunk 5]
```diff
@@ -179,7 +181,7 @@ public void init() throws DBException {
     String table = getProperties().getProperty(TABLENAME_PROPERTY, TABLENAME_PROPERTY_DEFAULT);
     try {
```
Comment (allanbank) [inline hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: If you merge the try above into this try you would have something like:

    if( THREAD_COUNT.getAndIncrement() == 0) {
        String table = getProperties().getProperty(TABLENAME_PROPERTY, TABLENAME_PROPERTY_DEFAULT);
        try {
            if (connection == null) {
                // Initialize if not set up already.
                connection = ConnectionFactory.createConnection(config);

                // Validate the table name.
                final TableName tName = TableName.valueOf(table);
                connection.getTable(tName).getTableDescriptor();
            }
        } catch (java.io.IOException e) {
            throw new DBException(e);
       }
    }

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java (hunk 1 above)
Comment (manolama) [reply hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: doh, thanks.

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java (hunk 5 above)
Comment (manolama) [reply hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: Good point, merged em.

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java (hunk 3 above)
Comment (manolama) [reply hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: Good point, yanked the sync.

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java (hunk 2 above)
Comment (manolama) [reply hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: He did mention it but I need confirmation from an HBase dev that the table fetch is thread safe in 1.X. I think it wasn't in the past.

Comment (busbey) [reply hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: Connection objects in hbase 1.y are threadsafe, so getTable is fine. The returned Table object is not threadsafe, but making them is supposed to be lightweight and done per-thread when needed.

**Comment by manolama** [comment]
Yanked the lock so we should be good now @busbey, thanks!

File: hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java [hunk 6]
```diff
@@ -208,15 +201,10 @@ public void cleanup() throws DBException {
...
-      if (threadCount.get() <= 0) {
-        // Means we are done so ok to shut down the Connection.
-        synchronized (CONNECTION_LOCK) {
-          if (connection != null) {
-            connection.close();
-            connection = null;
-          }
-        }
+      int threadCount = THREAD_COUNT.decrementAndGet();
+      if (threadCount <= 0 && connection != null) {
```
[the remark this answers is not in the document]
Comment (manolama) [reply hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: But Alan was right in that this should only fire when the atomic int returns a 0 at which point only one thread will perform the null check and close + null the connection.

Or are you thinking of the case where there may be some initialization issue wherein a run initializes multiple threads but the initializations fail and possibly cleanup is executed on a thread before initialization begins in another thread?

[the remark this answers is not in the document]
Comment (busbey) [reply hbase10/src/main/java/com/yahoo/ycsb/db/HBaseClient10.java]: Right exactly.

**Pull request merged**